File: python/features/steps/issue_key_steps.py
"""Behave steps for issue key representation."""

from __future__ import annotations

from behave import given, then, when

from taskulus.ids import format_issue_key


@given('an issue identifier "{identifier}"')
def given_issue_identifier(context: object, identifier: str) -> None:
    context.issue_identifier = identifier


@given('the display context is "{context_value}"')
def given_display_context(context: object, context_value: str) -> None:
    context.project_context = context_value == "project"


@when("I format the issue key")
def when_format_issue_key(context: object) -> None:
    identifier = getattr(context, "issue_identifier", "")
    project_context = getattr(context, "project_context", False)
    context.formatted_issue_key = format_issue_key(
        identifier=identifier,
        project_context=project_context,
    )


@then('the formatted key should be "{expected}"')
def then_formatted_key_should_match(context: object, expected: str) -> None:
    assert getattr(context, "formatted_issue_key", None) == expected

File: rust/features/steps/issue_key_steps.rs
use cucumber::{given, then, when};

use crate::steps::initialization_steps::TaskulusWorld;
use taskulus::ids::format_issue_key;

#[given(expr = "an issue identifier {string}")]
fn given_issue_identifier(world: &mut TaskulusWorld, identifier: String) {
    world.generated_id = Some(identifier);
}

#[given(expr = "the display context is {string}")]
fn given_display_context(world: &mut TaskulusWorld, context: String) {
    world.display_context = Some(context);
}

#[when("I format the issue key")]
fn when_format_issue_key(world: &mut TaskulusWorld) {
    let identifier = world
        .generated_id
        .as_ref()
        .cloned()
        .unwrap_or_else(|| "".to_string());
    let context = world
        .display_context
        .as_ref()
        .map(|value| value == "project")
        .unwrap_or(false);
    let formatted = format_issue_key(&identifier, context);
    world.formatted_issue_key = Some(formatted);
}

#[then(expr = "the formatted key should be {string}")]
fn then_formatted_key_should_match(world: &mut TaskulusWorld, expected: String) {
    let formatted = world
        .formatted_issue_key
        .as_ref()
        .expect("formatted key present");
    assert_eq!(formatted, &expected);
}
